src/wallet/rpcwallet.cpp

```diff
@@ -387,23 +387,22 @@ UniValue listaddresses(const UniValue& params, bool fHelp)
     // This will include any address that has been generated by this wallet.
     for (const std::pair<CTxDestination, CAddressBookData>& item : pwalletMain->mapAddressBook) {
         t_generated_dests.insert(item.first);
+        auto script = GetScriptForDestination(item.first);
+        if (pwalletMain->HaveWatchOnly(script)) {
+            t_watchonly_dests.insert(item.first);
+        }
     }
 
     // Ensure we have every address that holds a balance. While this is likely to be redundant
     // with respect to the entries in the address book for addresses generated by this wallet,
     // there is not a guarantee that an externally generated address (such as one associated with
     // a future unified incoming viewing key) will have been added to the address book.
     for (const std::pair<CTxDestination, CAmount>& item : pwalletMain->GetAddressBalances()) {
-        auto script = GetScriptForDestination(item.first);
-        if (pwalletMain->HaveWatchOnly(script)) {
-            t_watchonly_dests.insert(item.first);
-        } else if (t_generated_dests.count(item.first) == 0) {
+        if (t_generated_dests.count(item.first) == 0) {
```

Comment:
Why is this change being made?

Reply:
Because watch-only addresses are being added in the loop just above (over the address book), in other words, it's a move. I'm not sure this is correct, but without this move, an address had to have funds to be listed. This RPC should list addresses regardless of whether they contain funds, shouldn't it?

Reply:
Yes, we want to ideally see every address we want to expose (i.e. excluding unified change addresses). I think that was indeed a bug with the existing RPC, but I think there's also a bug in the fix (per my other comment). I'd have prefered that this be fixed separately to the unified changes in this PR, so they could be tested separately.

Reply:
We need to filter out any address that is associated with a UA here.

Reply:
```suggestion
        if (t_generated_dests.count(item.first) == 0 &&
            t_watchonly_dests.count(item.first) == 0 &&
            GetSourceForPaymentAddress(pwalletMain)(item.first) == PaymentAddressSource::Random)
        {
```
This is sufficient, because we'll have handled all manually-generated and imported addresses via the addressbook, and `GetSourceForPaymentAddress` ensures we filter out UA receivers.